src/material/schematics/migration-utilities/typescript/import-operations.ts

```diff
@@ -41,21 +32,32 @@ export function updateNamedImport(
     const n = node.elements[i];
     const name = n.propertyName ? n.propertyName : n.name;
     if (name.escapedText === opts.oldExport) {
-      return {
-        offset: name.pos,
-        updateFn: (text: string) => {
-          const index = text.indexOf(opts.oldExport, name.pos);
-          return replaceAt(text, index, {
-            old: opts.oldExport,
-            new: opts.newExport,
-          });
-        },
-      };
+      return createUpdate(name, {old: opts.oldExport, new: opts.newExport});
     }
   }
   return;
 }
 
+export function updateImportExpressionArgument(
+  node: ts.CallExpression,
+  opts: {
+    arg: string;
+  },
+): Update {
+  const arg = node.arguments[0] as ts.StringLiteral;
+  return createUpdate(arg, {old: arg.text, new: opts.arg});
+}
+
+function createUpdate(node: ts.Node, substring: {old: string; new: string}): Update {
```

Comment:
OOC: Why do we need to handle everything through captured updates, sorted by indices, if we could directly apply the updates? The original file indices are usable and that is what we use in all other migrations: https://github.com/angular/angular-cli/blob/3884b865262c1ffa5652ac0f4d67bbf59087f453/packages/angular_devkit/schematics/src/utility/update-buffer.ts#L330

just mentioning this for additional information. should simplify things

Reply:
I'm a little bit confused, but it seems like you're saying the devkit update recorder/magic-string already solves this problem in a standardized way? I've never tried using it, but from what I see here it looks promising

https://github.com/angular/angular-cli/blob/3884b865262c1ffa5652ac0f4d67bbf59087f453/packages/angular_devkit/schematics/src/utility/update-buffer_spec.ts#L335

I'll give it a shot and re-request review once I'm done

Reply:
yeah, feel free to land this for now, but I'm just trying to avoid these common schematics utilities as all other migrations never required this. The magic string/update-recorder really handles this "magically". This is also why the insert left and insert right methods exist -- trying to keep original indices as is -- and placing stuff in between without shifting offsets known to the recorder. Magic string has a good readme IIRC